[FIX] website_sale: forbid payment if an error is displayed

nda-odoo opened the pull request:
**steps to reproduce:**
- as admin, go to /shop and add a storable product to cart
- disable all delivery methods
- use a coupon to set the price to 0

**before this commit:**
- the payment button is clickable even if there is a big red error message saying "Sorry, we are unable to ship your order"

**after this commit:**
- the payment button is hidden if an error is displayed and the route /shop/payment/validate is blocked if there is an error displayed

opw-3582207-nda

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/142401).

vchu-odoo commented on addons/website_sale/controllers/main.py:
```diff
@@ -1131,6 +1131,12 @@ def shop_payment_validate(self, transaction_id=None, sale_order_id=None, **post)
             order = request.env['sale.order'].sudo().browse(sale_order_id)
             assert order.id == request.session.get('sale_last_order_id')
 
+        render_values = self._get_shop_payment_values(order, **post)
```
It is better to repeat only the check instead of redoing all the logic to create render values that are never used.
```
has_storable_products = any(line.product_id.type in ['consu', 'product'] for line in order.order_line)
if not order._get_delivery_methods() and has_storable_products:
     raise ValidationError(_('Unable to validate your order.'))
```

nda-odoo replied:
Hello! The thing is that the error can come from another module and I wanted to block it in a more generic way.

Ex:
https://github.com/odoo-dev/enterprise/blob/6bfa82866e86963050decea9841d2ef1c7a72185/website_sale_account_avatax/controllers/main.py#L9-L22

I could extract those errors in a submethod that i would call `_get_shop_payment_errors` in all those modules so I can use them without the render logic but the diff will be a bit bigger.

vchu-odoo replied:
Oh, you are right!
It would be great for the future to indeed do a bit of refactoring and extract the logic as you suggest.

vchu-odoo replied:
Yes, i think now it makes sense to separate your commit. I will help with resolving conflicts in 17 template. Could you please change the fw up to master @AntoineVDV ?

vchu-odoo commented on addons/website_sale/views/templates.xml:
```diff
@@ -1582,7 +1582,7 @@
                                 </a>
                             </div>
-                            <div class="js_payment mt-3" t-if="not website_sale_order.amount_total" id="payment_method" name="o_website_sale_free_cart">
+                            <div class="js_payment mt-3" t-if="not errors and (not website_sale_order.amount_total)" id="payment_method" name="o_website_sale_free_cart">
```
```suggestion
                            <div class="js_payment mt-3" t-if="not errors and not website_sale_order.amount_total" id="payment_method" name="o_website_sale_free_cart">
```
The brackets are unnecessary as the 'not' operation has higher priority.

vchu-odoo reviewed (changes requested):
Hello,

Thank you for your work on such an important issue!
You can see only some small suggestions from me.

vchu-odoo commented:
When this PR is merged please fw it only to **saas-16.4**. Since 17.0 templates were changed a lot, I will include this fix to 17 version in the following PR.

- https://github.com/odoo/odoo/pull/142265

AntoineVDV commented:
@fw-bot up to saas-16.4

vchu-odoo reviewed (changes requested):
Thank you for your changes, now it is much more clear!
Just small suggestions.

nda-odoo commented:
> @nda-odoo Pro tip: Never link a PR in a commit message because each push will contribute to adding noise to the linked PR. Instead, reference the task ID in the commit and use the PR's description to link another PR.

Thanks for the tip! I removed the PR from the enterprise commit message.

Thanks for the followup @vchu-odoo! I changed `+=` to `append` in both PR and added a docstring

vchu-odoo reviewed (approved):
LGTM, great work!
Final review from @AntoineVDV.

AntoineVDV commented on addons/website_sale/controllers/main.py:
```diff
@@ -1072,6 +1072,14 @@ def _get_shop_payment_values(self, order, **kwargs):
 
+    def _get_shop_payment_errors(self, order):
+        """ Check that there is no error that should block the payment.
+        :param order: The sales order to pay
```
```suggestion
        :param sale.order order: The sales order to pay
```

AntoineVDV commented on addons/website_sale/controllers/main.py:
```diff
@@ -1072,6 +1072,14 @@ def _get_shop_payment_values(self, order, **kwargs):
+        """ Check that there is no error that should block the payment.
+        :param order: The sales order to pay
+        :return: A list of errors (error_title, error_message)
+        :rtype: list(tuple)
```
```suggestion
        :rtype: list[tuple]
```

AntoineVDV commented on addons/website_sale_delivery/views/website_sale_delivery_templates.xml:
```diff
@@ -40,7 +40,7 @@
 
     <template id="payment_delivery" name="Delivery Costs" inherit_id="website_sale.payment">
-        <xpath expr="//t[@t-if='website_sale_order.amount_total']" position="before">
+        <xpath expr="//t[@name='website_sale_non_free_cart']" position="before">
```
You cannot xpath a node that you are defining at the same time in a different module in stable. If customers update the `website_sale_delivery` module but not `website_sale`, it will crash.

nda-odoo replied:
Sorry, I thought the stable policy didn't apply in that case. But I get it.

Sorry for the dumb question but what is be the "good" R&D way to handle this? I'm changing the "`t-if`" condition, so this xpath will break. Should I add a "or" in the xpath to keep the previous one ? Or use a different xpath selector that works on both, targeting the parent or the first child or whatever?

AntoineVDV replied:
Honestly, the only good answer is "whatever works" :grimacing: 

It can get as ugly as you like, as long as it's cleaned in the forward-port to master.

nda-odoo replied:
Great! I chose an ugly but simple

```
<!-- //t[@t-if='website_sale_order.amount_total'] should be removed in master -->
<xpath expr="//t[@name='website_sale_non_free_cart'] | //t[@t-if='website_sale_order.amount_total']" position="before">
```
and i will remove the `//t[@t-if='website_sale_order.amount_total']` part in master

AntoineVDV reviewed (changes requested):
@fw-bot up to master

AntoineVDV reviewed (approved):
@robodoo r+

robodoo commented:
@nda-odoo @AntoineVDV linked pull request(s) odoo/enterprise#51216 not ready. Linked PRs are not staged until all of them are ready.
